flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java

```diff
@@ -0,0 +1,235 @@
+	private CompletableFuture<DispatcherGateway> dispatcherGatewayFuture;
+
+	DefaultDispatcherRunner(
+			LeaderElectionService leaderElectionService,
+			FatalErrorHandler fatalErrorHandler,
+			DispatcherLeaderProcessFactory dispatcherLeaderProcessFactory) throws Exception {
+		this.leaderElectionService = leaderElectionService;
+		this.fatalErrorHandler = fatalErrorHandler;
+		this.dispatcherLeaderProcessFactory = dispatcherLeaderProcessFactory;
+		this.terminationFuture = new CompletableFuture<>();
+		this.shutDownFuture = new CompletableFuture<>();
+
+		this.isRunning = true;
+		this.dispatcherLeaderProcess = StoppedDispatcherLeaderProcess.INSTANCE;
+		this.previousDispatcherLeaderProcessTerminationFuture = CompletableFuture.completedFuture(null);
+		this.dispatcherGatewayFuture = new CompletableFuture<>();
+
+		startDispatcherRunner(leaderElectionService);
+	}
+
+	private void startDispatcherRunner(LeaderElectionService leaderElectionService) throws Exception {
+		LOG.info("Starting {}.", getClass().getName());
+
+		leaderElectionService.start(this);
+	}
+
+	@Override
+	public CompletableFuture<DispatcherGateway> getDispatcherGateway() {
+		synchronized (lock) {
+			return dispatcherGatewayFuture;
+		}
+	}
+
+	@Override
+	public CompletableFuture<ApplicationStatus> getShutDownFuture() {
+		return shutDownFuture;
+	}
+
+	@Override
+	public CompletableFuture<Void> closeAsync() {
+		synchronized (lock) {
+			if (!isRunning) {
+				return terminationFuture;
+			} else {
+				isRunning = false;
+			}
+		}
+
+		stopDispatcherLeaderProcess();
+		final CompletableFuture<Void> servicesTerminationFuture = stopServices();
+
+		FutureUtils.forward(
+			FutureUtils.completeAll(
+				Arrays.asList(
+					previousDispatcherLeaderProcessTerminationFuture,
+					servicesTerminationFuture)),
+			terminationFuture);
+
+		return terminationFuture;
+	}
+
+	private CompletableFuture<Void> stopServices() {
+		Exception exception = null;
+
+		try {
+			leaderElectionService.stop();
+		} catch (Exception e) {
+			exception = e;
+		}
+		if (exception == null) {
+			return CompletableFuture.completedFuture(null);
+		} else {
+			return FutureUtils.completedExceptionally(exception);
+		}
+	}
+
+	// ---------------------------------------------------------------
+	// Leader election
+	// ---------------------------------------------------------------
+
+	@Override
+	public void grantLeadership(UUID leaderSessionID) {
+		runActionIfRunning(() -> startNewDispatcherLeaderProcess(leaderSessionID));
+	}
+
+	private void startNewDispatcherLeaderProcess(UUID leaderSessionID) {
+		stopDispatcherLeaderProcess();
+
+		createAndAssignNewDispatcherLeaderProcess(leaderSessionID);
+
+		final DispatcherLeaderProcess newDispatcherLeaderProcess = dispatcherLeaderProcess;
+		FutureUtils.assertNoException(
+			previousDispatcherLeaderProcessTerminationFuture.thenRun(newDispatcherLeaderProcess::start));
+	}
+
+	private void createAndAssignNewDispatcherLeaderProcess(UUID leaderSessionID) {
```

Comment:
Redundant and hard to reason due to its `void` signature. "Assign" is already implied by `=`. Could be inlined.

Reply:
True, this method can be removed. Will do it.